opencensus/common/schedule/__init__.py

```diff
@@ -50,3 +52,121 @@ def run(self):
+        return ('{}({})'.format(type(self).__name__, self.name))
+
+    def set(self):
+        return self.event.set()
+
+    def wait(self, timeout=None):
+        return self.event.wait(timeout)
+
+
+class QueueExitEvent(QueueEvent):
+    pass
+
+
+class Queue(queue.Queue):
+    def __init__(self, capacity):
+        self.EXIT_EVENT = QueueExitEvent('EXIT')
+        self._queue = queue.Queue(maxsize=capacity)
+
+    def _gets(self, count, timeout):
+        start_time = time.time()
+        elapsed_time = 0
+        cnt = 0
+        while cnt < count:
+            try:
+                item = self._queue.get(block=False)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+        while cnt < count:
+            wait_time = max(timeout - elapsed_time, 0)
+            try:
+                item = self._queue.get(block=True, timeout=wait_time)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+            elapsed_time = time.time() - start_time
+
+    def gets(self, count, timeout):
+        return tuple(self._gets(count, timeout))
+
+    def flush(self, timeout=None):
+        start_time = time.time()
+        wait_time = timeout
+        event = QueueEvent('SYNC(timeout={})'.format(wait_time))
+        try:
+            self._queue.put(event, block=True, timeout=wait_time)
+        except queue.Full:
+            return
+        elapsed_time = time.time() - start_time
+        wait_time = timeout and max(timeout - elapsed_time, 0)
+        if event.wait(timeout):
+            return time.time() - start_time  # time taken to flush
+
+    def put(self, item, block=True, timeout=None):
+        try:
+            self._queue.put(item, block, timeout)
+        except queue.Full:
+            pass  # TODO: log data loss
+
+    def puts(self, items, block=True, timeout=None):
+        if block and timeout is not None:
+            start_time = time.time()
+            elapsed_time = 0
+            for item in items:
+                wait_time = max(timeout - elapsed_time, 0)
+                self.put(item, block=True, timeout=wait_time)
+                elapsed_time = time.time() - start_time
+        else:
+            for item in items:
+                self.put(item, block, timeout)
+
+
+class Worker(threading.Thread):
+    daemon = True
+
+    def __init__(self, src, dst):
+        self.src = src
+        self.dst = dst
+        self._stopping = False
+        super(Worker, self).__init__()
+
+    def run(self):
+        src = self.src
+        dst = self.dst
+        while True:
+            batch = src.gets(dst.max_batch_size, dst.export_interval)
+            if batch and isinstance(batch[-1], QueueEvent):
+                dst.emit(batch[:-1], event=batch[-1])
```

Comment:
I was thinking this as a benefit to tell the exporter explicitly about the intention.
For example, if we have a slow network, the exporter can decide to persist the data locally to prevent data loss on exit / flush.

Reply:
There are two possible ways in my mind:
1. Having a base exporter that handle the event by default.

```python

class BaseExporter(object):
    def export_internal(self, batch, event):
        try:
            return self.export(batch)
        finally:
            if event:
                event.set()

    def export(self, batch):
        pass
```

2. Use runtime inspection (a bit dirty).

```python
export_thunk = exporter.export
if 'event' in inspect.signature(export_thunk).parameters:
    def export(batch, event):
        try:
            exporter.export(batch)
        finally:
            if event:
                event.set()
    export_thunk = export
```

Reply:
(1) looks better to me, but does `shutdown` solve the same problem this does?